### ISSUE TITLE
Custom "Recurrent" Wrapper layer Verification

### ISSUE THREAD

**Issue by Junyoungpark** [keras-team/keras#7815]
Hi.

I'm recently implementing a Keras layer for modeling recurrent neural network with arbitrary neural network based-cell. As an example, I would like to train a network for approximating following dynamic system equation.

<a href="https://www.codecogs.com/eqnedit.php?latex=s_t_&plus;_1&space;=&space;f(s_t,&space;a_t)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_t_&plus;_1&space;=&space;f(s_t,&space;a_t)" title="s_t_+_1 = f(s_t, a_t)" /></a>

where <a href="https://www.codecogs.com/eqnedit.php?latex=s_t" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_t" title="s_t" /></a> is a state variable at time t and <a href="https://www.codecogs.com/eqnedit.php?latex=a_t" target="_blank"><img src="https://latex.codecogs.com/gif.latex?a_t" title="a_t" /></a> is an action variable at time t. Also, <a href="https://www.codecogs.com/eqnedit.php?latex=f" target="_blank"><img src="https://latex.codecogs.com/gif.latex?f" title="f" /></a> is a function what I like to approximate with  deep learning.

As a simple approach, One may design a feed forward neural network, let say NN, that takes <a href="https://www.codecogs.com/eqnedit.php?latex=s_t" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_t" title="s_t" /></a> and <a href="https://www.codecogs.com/eqnedit.php?latex=a_t" target="_blank"><img src="https://latex.codecogs.com/gif.latex?a_t" title="a_t" /></a> as inputs and train to predict <a href="https://www.codecogs.com/eqnedit.php?latex=s_t_&plus;_1" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s_t_&plus;_1" title="s_t_+_1" /></a> via back propagation. However, I would like to impose one more constraint to NN  by recurrently connecting itself. I conceptually denotes the recurrent connections in the following diagram.

![cdn](https://user-images.githubusercontent.com/3063343/30025150-1420004e-91b2-11e7-854d-5e0b8f22eea8.png)

Here are questions incoming. I implemented the recurrent connections with a Keras layer that slightly modify `TimeDistributed` layer. Firstly, I modified and restored recurrent connection for estimated state by modifying `step` method of `TimeDistributed` layer.

Following is full implementation of my poor code.

```
class RecurrentWrapper(TimeDistributed):
    
    def build(self, input_shape):
        #input_shape = [(None, 1 (=initial State), state Dimension), (none, TimeSteps, actionDim)]
        
        #self.state_input_spec = InputSpec(shape=input_shape[0])
        #self.action_input_spec = InputSpec(shape=input_shape[1])
        
        BatchSisze = input_shape[0][0]
        clildInputDim = int(input_shape[0][-1]+input_shape[1][-1])        
        child_input_shape = (BatchSisze,) + (clildInputDim,)
        if not self.layer.built:
            self.layer.build(child_input_shape)
            self.layer.built = True
        super(TimeDistributed, self).build()
    
    def call(self, inputs, training=None, mask=None):
        """
            Args:
                inputs : 
                    a list of tensors. Assume first element is an action tensor with shape 
                    [Bacthsize, timeStep, #. of controll vars] and the second element is a stateTensor
                    with shape of [Batchsize, Dim of states]. For example, inputs = [actionPlaceHolder, statePlaceHolder].
        """
        kwargs = {}
        if has_arg(self.layer.call, 'training'):
            kwargs['training'] = training
        uses_learning_phase = False
        
        def _step(actions, states):
            concatStateAction = tf.concat([actions,states], axis=1)
            new_state = self.layer.call(concatStateAction)
            output = new_state
            return output, new_state
        
        initialState = inputs[0]
        actionProfile = inputs[1]
        lastOutput, outputs, states = K.rnn(_step, actionProfile, initialState, unroll=True, constants=0.0)
        
        return outputs # For this case output == states
    
    def compute_output_shape(self, input_shape):
        #input_shape = [(None, 1 (=initial State), state Dimension), (none, TimeSteps, actionDim)]
        
        BatchSisze = input_shape[0][0]
        clildInputDim = int(input_shape[0][-1]+input_shape[1][-1])        
        timesteps = input_shape[1][1]
        child_input_shape = (BatchSisze,) + (clildInputDim,)
        
        child_output_shape = self.layer.compute_output_shape(child_input_shape)
        
        return (child_output_shape[0], timesteps) + child_output_shape[1:]
```

Also, this is my usage for the `RecurrentWrapper`

```
stateInput = Input(shape=(dimState,), name='stateInput')
actionInput = Input(shape=(timeStep, dimAction), name='acitonInput')


NN = Sequential()
NN.add(Dense(64, activation='relu', input_shape=(dimState+dimAction,)))
NN.add(Dense(64, activation='relu'))
NN.add(Dense(64, activation='relu'))
NN.add(Dense(dimState))

RDN = RecurrentWrapper(NN)([stateInput,actionInput])
Rmodel = Model(inputs=[stateInput,actionInput],outputs=[RDN])
```

However, I'm stuck in problem. Since the prediction performance of the recurrent network is significantly poor than that of NN. 

Therefore, **I'd like to verify my implementation coincides with the recurrent concept**. And **loss update has properly happened.** And finally, **Graph connection is appropriately constructed or not**.
How can I figure out the problems.

Thanks in advance.
Junyoung Park.
